_deploy/r/gnoswap/gnft/gnft.gno

```diff
@@ -148,75 +197,121 @@ func Render(path string) string {
-	r := rand.New(pcg)
-
-	tokenURI := genImageURI(r)
-	ok, _ := gnft.SetTokenURI(tid, grc721.TokenURI(tokenURI))
-	if ok {
-		prevAddr, prevRealm := getPrev()
-		std.Emit(
-			"SetTokenURI",
-			"prevAddr", prevAddr,
-			"prevRealm", prevRealm,
-			"lpTokenId", string(tid),
-			"internal_tokenURI", tokenURI,
-		)
+	caller := std.PrevRealm().Addr()
+	tokenListByCaller, _ := getTokenList(caller)
+	lenTokenListByCaller := len(tokenListByCaller)
+	if lenTokenListByCaller == 0 {
+		return nil
 	}
-}
 
-func SetTokenURILast() {
-	common.IsHalted()
+	lastToken := tokenListByCaller[lenTokenListByCaller-1]
+	tokenURI, err := SetTokenURI(lastToken)
+	if err != nil {
+		return err
+	}
+
+	prevAddr, prevPkgPath := getPrevAsString()
+	std.Emit(
+		"SetTokenURILast",
+		"prevAddr", prevAddr,
+		"prevRealm", prevPkgPath,
+		"lpTokenId", string(lastToken),
+		"internal_tokenURI", tokenURI,
+	)
 
-	// rand instance
+	return nil
+}
+
+// generateRandInstnace generates a new random instance
+// Returns:
+//   - *rand.Rand: A new random instance
+func generateRandInstance() *rand.Rand {
 	seed1 := uint64(time.Now().Unix())
 	seed2 := uint64(time.Now().UnixNano())
 	pcg := rand.NewPCG(seed1, seed2)
-	r := rand.New(pcg)
-
-	prev := std.PrevRealm()
-	prevAddr := prev.Addr()
-	usersNFT := userMinted[prevAddr]
-	for _, tid := range usersNFT {
-		tokenURI := genImageURI(r)
-
-		ok, _ := gnft.SetTokenURI(tid, grc721.TokenURI(tokenURI))
-		if ok {
-			prevAddr, prevRealm := getPrev()
-			std.Emit(
-				"SetTokenURI",
-				"prevAddr", prevAddr,
-				"prevRealm", prevRealm,
-				"lpTokenId", string(tid),
-				"internal_tokenURI", tokenURI,
-			)
-		}
+	return rand.New(pcg)
+}
+
+// getTokenList retrieves the list of nft tokens for an address
+// Parameters:
+//   - addr: The address to check for nft tokens
+//
+// Returns:
+//   - []grc721.TokenID: Array of token IDs
+//   - bool: true if tokens exist for the address, false otherwise
+func getTokenList(addr std.Address) ([]grc721.TokenID, bool) {
+	iTokens, exists := tokenList.Get(addr.String())
+	if !exists {
+		return []grc721.TokenID{}, false
+	}
+
+	return iTokens.([]grc721.TokenID), true
+}
+
+// mustGetTokenList same as getTokenList but panics if tokens don't exist
+// Parameters:
+//   - addr: The address to check for nft tokens
+//
+// Returns:
+//   - []grc721.TokenID: Array of token IDs
+func mustGetTokenList(addr std.Address) []grc721.TokenID {
+	tokens, exists := getTokenList(addr)
+	if !exists {
+		panic(ufmt.Sprintf("user %s has no minted nft tokens", addr.String()))
 	}
 
-	delete(userMinted, prevAddr)
+	return tokens
+}
+
+// appendTokenList adds a token ID to the list of nft tokens
+// Parameters:
+//   - addr: The address to append the token for
+//   - tid: The token ID to append
+func appendTokenList(addr std.Address, tid grc721.TokenID) {
+	prevTokenList, _ := getTokenList(addr)
+	prevTokenList = append(prevTokenList, tid)
+	tokenList.Set(addr.String(), prevTokenList)
+}
+
```

Comment:
The removeTokenList function is added and should be managed for changes in the asset.

Reply:
00756c7d9cacc5d7892c8c8266788d980d6f4c4c